capsules/src/net/lowpan_fragment.rs

```diff
@@ -0,0 +1,900 @@
+            payload_len
+        };
+        self.packet.replace(packet);
+        if !self.bitmap.map_or(false, |bitmap| {
+            bitmap.set_bits(dgram_offset / 8, (dgram_offset + uncompressed_len) / 8)
+        }) {
+            // If this fails, we received an overlapping fragment. We can simply
+            // drop the packet in this case.
+            Err(ReturnCode::FAIL)
+        } else {
+            self.bitmap
+                .map(|bitmap| bitmap.is_complete((dgram_size as usize) / 8))
+                .ok_or(ReturnCode::FAIL)
+        }
+    }
+
+    fn end_receive(&self, client: Option<&'static ReceiveClient>, result: ReturnCode) {
+        self.busy.set(false);
+        self.bitmap.map(|bitmap| bitmap.clear());
+        self.timeout_counter.set(0);
+        client.map(move |client| {
+            // Since packet is borrowed from the upper layer, failing to return it
+            // in the callback represents a significant error that should never
+            // occur - all other calls to `packet.take()` replace the packet,
+            // and thus the packet should always be here.
+            let mut buffer =
+                self.packet.take().expect("Error: `packet` is None in call to end_receive.");
+            client.receive(&buffer, self.dgram_size.get(), result);
+            self.packet.replace(buffer);
+        });
+    }
+}
+
+/// struct FragState
+/// ----------------
+/// This struct tracks the global sending/receiving state, and contains the
+/// lists of RxStates and TxStates.
+pub struct FragState<'a, A: time::Alarm + 'a> {
+    pub radio: &'a Mac,
+    ctx_store: &'a ContextStore,
+    alarm: &'a A,
+
+    // Transmit state
+    tx_states: List<'a, TxState<'a>>,
+    tx_dgram_tag: Cell<u16>,
+    tx_busy: Cell<bool>,
+    tx_buf: TakeCell<'static, [u8]>,
+
+    // Receive state
+    rx_states: List<'a, RxState<'a>>,
+    rx_client: Cell<Option<&'static ReceiveClient>>,
+}
+
+// This function is called after transmitting a frame
+#[allow(unused_must_use)]
+impl<'a, A: time::Alarm> TxClient for FragState<'a, A> {
+    fn send_done(&self, buf: &'static mut [u8], acked: bool, result: ReturnCode) {
+        self.tx_buf.replace(buf);
+        if result != ReturnCode::SUCCESS {
+            self.end_packet_transmit(acked, result);
+            return;
+        }
+        self.tx_states.head().map(move |head| {
+            if head.is_transmit_done() {
+                // This must return Some if we are in the closure - in particular,
+                // tx_state == head
+                self.end_packet_transmit(acked, result);
+            } else {
+                // Otherwise, we found an error
+                // Note that `tx_buf` should *always* be here, as we called
+                // `self.tx_buf.replace(..)` at the top of this function.
+                let tx_buf =
+                    self.tx_buf.take().expect("Error: `tx_buf` is None in send_done callback.");
```

Comment:
The `expect` call on line 582 is for clarity; we replace the `tx_buf` buffer on line 567, so there is no way for this call to fail. Although the `expect` call could be removed by only calling `self.tx_buf.replace(..)` on control paths where `head.prepare_transmit_next_fragment` doesn't succeed, I believe this is the cleanest way to write this code.